src/webui/api/searchcontroller.cpp

```diff
@@ -0,0 +1,286 @@
+
+class SearchPluginManager;
+
+using Utils::String::parseBool;
+
+void SearchController::startAction()
+{
+    checkParams({"pattern", "category", "plugins"});
+
+    if (Utils::Misc::pythonVersion() < 0)
+        throw APIError(APIErrorType::Conflict, "Python must be installed to use the Search Eengine.");
+
+    const QString pattern = params()["pattern"].trimmed();
+    const QString category = params()["category"].trimmed();
+    const QString plugins = params()["plugins"].trimmed();
+    const QString pluginsLower = plugins.toLower();
+
+    QStringList pluginsToUse;
+    if (pluginsLower == "all") pluginsToUse = SearchPluginManager::instance()->allPlugins();
+    else if (pluginsLower == "enabled") pluginsToUse = SearchPluginManager::instance()->enabledPlugins();
+    else if (pluginsLower == "multi") pluginsToUse = SearchPluginManager::instance()->enabledPlugins();
+    else pluginsToUse << plugins;
+
+    const QString id = generateSearchId();
+    auto *searchHandler = SearchPluginManager::instance()->startSearch(pattern, category, pluginsToUse);
+    m_searches[id] = new WebSearchHandler(searchHandler);
+
+    QVariantMap resultMap;
+    resultMap.insert("status", "Ok.");
+    resultMap.insert("id", id);
+
+    setResult(QJsonObject::fromVariantMap(resultMap));
+}
+
+void SearchController::stopAction()
+{
+    checkParams({"id"});
+
+    const QString id = params()["id"].trimmed();
+    if (!m_searches.contains(id)) {
+        throw APIError(APIErrorType::BadParams);
+    }
+
+    WebSearchHandler *webSearchHandler = m_searches[id];
+
+    if (!webSearchHandler->isActive()) {
+        setResult("Nothing to stop.");
+        return;
+    }
+
+    webSearchHandler->cancelSearch();
+    m_searches.remove(id);
+
+    setResult("Ok.");
+}
+
+void SearchController::resultsAction()
+{
+    checkParams({"id"});
+
+    const QString id = params()["id"].trimmed();
+    if (!m_searches.contains(id)) {
+        setResult(QJsonObject::fromVariantMap(getResults(QList<SearchResult>(), false, 0)));
+        return;
+    }
+
+    WebSearchHandler *webSearchHandler = m_searches[id];
+    const QList<SearchResult> searchResults = webSearchHandler->getSearchResults();
+
+    setResult(QJsonObject::fromVariantMap(getResults(searchResults, webSearchHandler->isActive(), webSearchHandler->getQueueSize())));
+
+    if (!webSearchHandler->isActive())
+        m_searches.remove(id);
+}
+
+void SearchController::downloadTorrent()
+{
+    checkParams({"id", "torrentUrl", "siteUrl"});
+
+    const QString id = params()["id"].trimmed();
+    const QString torrentUrl = params()["torrentUrl"].trimmed();
+    const QString siteUrl = params()["siteUrl"].trimmed();
+
+    WebSearchHandler *webSearchHandler = m_searches[id];
+    if (!m_searches.contains(id)) {
+        throw APIError(APIErrorType::BadParams);
+    }
+
+    webSearchHandler->downloadTorrent(torrentUrl, siteUrl);
+}
+
+void SearchController::categoriesAction()
+{
+    QStringList categories;
```

Comment:
QSet<QString>?

Reply:
QSet doesn't maintain order of its elements, which is required for categories.